File: MtxVec.Delphi/SysLinear1.pas
unit SysLinear1;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  Basic2, StdCtrls, ExtCtrls,MtxVecEdit,MtxVec, Grids, ComCtrls, MtxExpr;

type
  TLinearSystem1 = class(TBasicForm2)
    StringGrid1: TStringGrid;
    Label1: TLabel;
    StringGrid2: TStringGrid;                 
    Label2: TLabel;
    Label3: TLabel;
    StringGrid3: TStringGrid;
    Label4: TLabel;
    Label5: TLabel;
    Button1: TButton;
    Panel4: TPanel;
    Memo2: TMemo;
    RadioGroup1: TRadioGroup;
    CheckBox1: TCheckBox;
    Label6: TLabel;
    ComboBox1: TComboBox;
    Panel5: TPanel;
    Label7: TLabel;
    StringGrid4: TStringGrid;
    Label8: TLabel;
    procedure FormCreate(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure CheckBox1Click(Sender: TObject);
    procedure RadioGroup1Click(Sender: TObject);
    procedure ComboBox1Change(Sender: TObject);
  private
    A: Matrix;
    b,x,S : Vector;
    { Private declarations }
  public
    { Public declarations }
  end;

var
  LinearSystem1: TLinearSystem1;

implementation

Uses MtxDialogs;

{$IFDEF CLR}
{$R *.NFM}
{$ELSE}
{$R *.DFM}
{$ENDIF}

procedure TLinearSystem1.FormCreate(Sender: TObject);
begin
  inherited;
  With RichEdit1.Lines, RichEdit1 do
  begin
    Clear;
    Add('MtxVec offers three different methods to solve a system '
      + 'of linear equations : LU, LQR and singular value '
      + 'decomposition. Each of these methods has its advantages '
      + 'and disadvantages. Try changing values in matrix A, '
      + 'vector b and method used to calculate the solution vector '
      + 'x. The displayed matrix is nearly singular. You can add '
      + 'more non-zero elements to the diagonal to make it non-singular.');
    Add('In practice, most computations are performed with rounding '
      + 'errors. The LUSolve method offers you additional tools for '
      + 'refining the solution and estimating its error. They are '
      + 'evaluated only if RefineSolution is set to true');
    Add('');
    SelAttributes.Style := [fsUnderline];
    Add('Parameters in this demo :');
    Add('ConditionNr: If your matrix is ill-conditioned '
      + '(that is, it''s ConditionNr is very small), then the error in '
      + 'the solution x is also large. ');
    Add('BackError: the smallest relative perturbation in elements '
      + 'of A and b such that x is the exact solution of the perturbed '
      + 'system.');
    Add('ForwError: the component-wise forward error in the computed '
      + 'solution.');
    Add('');
    Add('Try changing the matrix values and observe the solution, CondtionNr, ForwError and BackError.');
  end;
  A.SetIt(4,4,false,[0.001,  0,     0, 0,
                     3    ,0.001,   0, 0,
                     2.5  ,  2,     3, 0,
                     4    , -1,   0.5, 0.1]);
  b.Size(4);
  b.Setit([3,1.5,-2,4]);
  X.Size(0);
  S.Size(0);
  ValuesToGrid(A,StringGrid1,0,0,'0.0000',false);
  ValuesToGrid(b,StringGrid3,0,0,'0.0000',false);
  ComboBox1.ItemIndex := 0;
  ComboBox1Change(ComboBox1);
  CheckBox1Click(CheckBox1);
  RadioGroup1Click(RadioGroup1);
end;

procedure TLinearSystem1.Button1Click(Sender: TObject);
begin
  GridToValues(A,StringGrid1,A.Complex,false);
  GridToValues(b,StringGrid3,b.Complex,true,false);
  Case ComboBox1.ItemIndex of
  0:begin
      A.LUSolve(b,x);
      Memo2.Lines.Clear;
      if CheckBox1.Checked and (RadioGroup1.ItemIndex <> 0) then
      begin
           Memo2.Lines.Add('Forward error = '+ FormatFloat('0.0000E+00 ',A.ForwError));
           Memo2.Lines.Add('Backward error = '+ FormatFloat('0.0000E+00 ',A.BackError));
           Memo2.Lines.Add('ConditionNr = '+ FormatFloat('0.0000E+00',A.ConditionNr));
      end;
    end;
  1:begin
      A.SVDSolve(b,x,S);
      ValuesToGrid(S,StringGrid4,0,0,'0.0000E+00',false);
    end;
  end;
  ValuesToGrid(x,StringGrid2,0,0,'0.0000E+00',false);
end;

procedure TLinearSystem1.CheckBox1Click(Sender: TObject);
begin
  A.RefineSolution := TCheckBox(Sender).Checked;
end;

procedure TLinearSystem1.RadioGroup1Click(Sender: TObject);
begin
     Case TRadioGroup(Sender).ItemIndex of
       0:   A.ConditionNumber := cnNone;
       1:   A.ConditionNumber := cnNorm1;
       2:   A.ConditionNumber := cnNormInf;
     end;
end;

procedure TLinearSystem1.ComboBox1Change(Sender: TObject);
begin
  Panel4.Visible := TComboBox(Sender).ItemIndex = 0;
  Panel5.Visible := TComboBox(Sender).ItemIndex = 1;
  Memo2.Lines.Clear;
end;

initialization
   RegisterClass(TLinearSystem1);

end.

File: MtxVec.CPP/SysLinear1.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "SysLinear1.h"
#include "MtxVecEdit.hpp"
#include "MtxExpr.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Basic2"
#pragma resource "*.dfm"
TLinearSystem1 *LinearSystem1;
//---------------------------------------------------------------------------
__fastcall TLinearSystem1::TLinearSystem1(TComponent* Owner)
  : TBasicForm2(Owner)
{
  AnsiString str;

  RichEdit1->Lines->Clear();
  str = "MtxVec offers three different methods to solve a system ";
  str+= "of linear equations : LU, LQR and singular value ";
  str+= "decomposition. Each of these methods has its advantages ";
  str+= "and disadvantages. Try changing values in matrix A, ";
  str+= "vector b and method used to calculate the solution vector ";
  str+= "x. The displayed matrix is nearly singular. You can add ";
  str+= "more non-zero elements to the diagonal to make it non-singular.";
  RichEdit1->Lines->Add(str);
  str = "In practice, most computations are performed with rounding ";
  str+= "errors. The LUSolve method offers you additional tools for ";
  str+= "refining the solution and estimating its error. They are ";
  str+= "evaluated only if RefineSolution is set to true";
  RichEdit1->Lines->Add(str);
  RichEdit1->Lines->Add("");
  RichEdit1->SelAttributes->Style = RichEdit1->SelAttributes->Style << TFontStyle::fsUnderline;
  RichEdit1->Lines->Add("Parameters in this demo :");
  str = "ConditionNr: If your matrix is ill-conditioned ";
  str+= "(that is, it""s ConditionNr is very small), then the error in ";
  str+= "the solution x is also large. ";
  RichEdit1->Lines->Add(str);
  str = "BackError: the smallest relative perturbation in elements ";
  str+= "of A and b such that x is the exact solution of the perturbed system.";
  RichEdit1->Lines->Add(str);
  RichEdit1->Lines->Add("ForwError: the component-wise forward error in the computed solution.");
  RichEdit1->Lines->Add("");

  RichEdit1->Lines->Add("Try changing the matrix values and observe the solution, CondtionNr, ForwError and BackError.");

  A = new Mtxvec::TMatrix;
  A->SetIt(4,4,false,OPENARRAY(double,(0.001,  0,   0,   0,
										3,      0.1,   0,   0,
										2.5,    2,   3,   0,
										4,     -1, 0.5, 0.1)));
  b = new Mtxvec::TVector;
  b->Size(4);
  b->SetIt(OPENARRAY(double,(3,1.5,-2,4)));
  ValuesToGrid(A, StringGrid1,0,0,"0.0000",false);
  ValuesToGrid(b,StringGrid3,0,0,"0.0000",false,true);
  ComboBox1->ItemIndex = 0;
  ComboBox1Change(ComboBox1);
  CheckBox1Click(CheckBox1);
  RadioGroup1Click(RadioGroup1);
}
//---------------------------------------------------------------------------

void __fastcall TLinearSystem1::Button1Click(TObject *Sender)
{
  sVector x, S;

  GridToValues(A, StringGrid1,A->Complex, false);
  GridToValues(b, StringGrid3,b->Complex,true,false);
  switch (ComboBox1->ItemIndex)
  {
    case 0:
		  {
//            A->LUSolve(b,x,mtGeneral,opNone);
			x = DivLeft(A,b);  // x = A^(-1)*b
            Memo2->Lines->Clear();
            if (CheckBox1->Checked && RadioGroup1->ItemIndex != 0)
            {
              Memo2->Lines->Add("Forward error = "+ FormatFloat("0.0000E+00 ",A->ForwError));
              Memo2->Lines->Add("Backward error = "+ FormatFloat("0.0000E+00 ",A->BackError));
              Memo2->Lines->Add("ConditionNr = "+ FormatFloat("0.0000E+00",A->ConditionNr));
            }
          };break;
    case 1:
          {
            A->SVDSolve(b,x,S,1.0E-9);
            ValuesToGrid(S, StringGrid4,0,0,"0.0000E+00",false,true);
          };break;
    }
  ValuesToGrid(x, StringGrid2,0,0,"0.0000E+00",false,true);
}
//---------------------------------------------------------------------------

void __fastcall TLinearSystem1::CheckBox1Click(TObject *Sender)
{
  A->RefineSolution = CheckBox1->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TLinearSystem1::RadioGroup1Click(TObject *Sender)
{
  switch (RadioGroup1->ItemIndex)
  {
    case 0: A->ConditionNumber = cnNone; break;
    case 1: A->ConditionNumber = cnNorm1; break;
    case 2: A->ConditionNumber = cnNormInf; break;
  }
}
//---------------------------------------------------------------------------

void __fastcall TLinearSystem1::FormDestroy(TObject *Sender)
{
  delete A;
  delete b;
}
//---------------------------------------------------------------------------

void __fastcall TLinearSystem1::ComboBox1Change(TObject *Sender)
{
  Panel4->Visible = (ComboBox1->ItemIndex == 0);
  Panel5->Visible = (ComboBox1->ItemIndex == 1);
  Memo2->Lines->Clear();
}
//---------------------------------------------------------------------------

